### PR TITLE
Anway shmget syscall

### PR DESCRIPTION
**Pull request by Anway-Agte**
## Description
This PR adds comments for the shmget_syscall which deals with shared memory mapping

### Type of change
- [ ] This change  adds documentation

## How Has This Been Tested?

- Test A - `ut_lind_test_shmget_syscall`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -4979,27 +4979,63 @@ impl Cage {
         None
     }
 
-    //------------------SHMGET SYSCALL------------------
-
-    pub fn shmget_syscall(&self, key: i32, size: usize, shmflg: i32) -> i32 {
+    /// ### Description
+    /// 
+    /// `shmget_syscall` returns the shared memory segment identifier associated with a particular `key`
+    /// If a key doesn't exist or equals `IPC_PRIVATE`, shmget creates a new
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs]: Can you explain a bit about `IPC_PRIVATE`.

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -4979,27 +4979,63 @@ impl Cage {
...
+    /// memory segment and attaches it to the key, however we only create a new memory segment if the key 
+    /// doesn't exist, and return an error if key equals `IPC_PRIVATE`
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs]: In the above statement, its mentioned that a new memory segment is created if the key doesn't exist or equals to IPC_PRIVATE. But, in the statement just below it, we are stating that if the key doesn't exist, only then we are creating a new memory segment. Looks like both statements are contradicting each other here. 
Can you check this once?

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -4979,27 +4979,63 @@ impl Cage {
...
         let shmid: i32;
+        // data of the shm table
         let metadata = &SHM_METADATA;
 
+        // Check if there exists a memory segment associated with the key passed as argument
         match metadata.shmkeyidtable.entry(key) {
+            // If there exists a memory segment at that key
             interface::RustHashEntry::Occupied(occupied) => {
+                // Produce an error if invalid flags are used with a valid key
                 if (IPC_CREAT | IPC_EXCL) == (shmflg & (IPC_CREAT | IPC_EXCL)) {
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs:5032]: Explaining a bit about both these flags and their usecases (`IPC_CREAT` and `IPC_EXCL`) would be great.

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -5008,6 +5044,9 @@ impl Cage {
                     );
                 }
 
+                // If memory segment doesn't exist and IPC_CREAT was specified - we create a new memory segment
+
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs]: Extra line (can be removed)

File: src/tests/fs_tests.rs [hunk 5]
```diff
@@ -4034,6 +4034,34 @@ pub mod fs_tests {
...
+        let cage = interface::cagetable_getref(1); 
+
+        let key = 33123;
+        // Get shmid of a memory segment / create a new one if it doesn't exist
+        let shmid = cage.shmget_syscall(33123, 1024, IPC_CREAT);       
+        assert_eq!(shmid,4); 
+
+        // Check error upon asking for a valid key and passing the IPC_CREAT or IPC_EXCL flag
+        assert_eq!(cage.shmget_syscall(key, 1024, IPC_CREAT),-(Errno::EEXIST as i32 ));
+        // assert_eq!(cage.shmget_syscall(key, 1024, IPC_EXCL),-(Errno::EEXIST as i32));
```
Comment (namanlalitnyu) [inline src/tests/fs_tests.rs]: I think this should be uncommented for testing the `IPC_EXCL` scenario.

**Review by namanlalitnyu — COMMENTED** [review]
Some minor comments.

File: src/safeposix/syscalls/fs_calls.rs (hunk 1 above)
Comment (Anway-Agte) [reply src/safeposix/syscalls/fs_calls.rs]: Added comments - please take a look!

File: src/tests/fs_tests.rs [hunk 6]
```diff
@@ -4034,6 +4034,33 @@ pub mod fs_tests {
...
+        // acquire locks and start env cleanup
+        let _thelock = setup::lock_and_init();
+        let cage = interface::cagetable_getref(1); 
+
+        let key = 33123;
+        // Get shmid of a memory segment / create a new one if it doesn't exist
+        let shmid = cage.shmget_syscall(33123, 1024, IPC_CREAT);       
+        assert_eq!(shmid,4); 
+
+        // Check error upon asking for a valid key and passing the IPC_CREAT or IPC_EXCL flag
```
Comment (ve1nard) [inline src/tests/fs_tests.rs]: you are passing both flags, not one of them, so I think "IPC_CREAT or IPC_EXCL flag" should be changed to "both IPC_CREAT and IPC_EXCL flags" for clarity.

File: src/tests/fs_tests.rs [hunk 7]
```diff
@@ -4034,6 +4034,33 @@ pub mod fs_tests {
...
+
+        lindrustfinalize();
+    }
+
+    pub fn ut_lind_fs_shmget_syscall(){
+        // acquire locks and start env cleanup
+        let _thelock = setup::lock_and_init();
+        let cage = interface::cagetable_getref(1); 
+
+        let key = 33123;
```
Comment (ve1nard) [inline src/tests/fs_tests.rs:4046]: can we add a simple test case for passing IPC_PRIVATE?

**Review by ve1nard — COMMENTED** [review]
great job, a couple of minor changes requested

File: src/tests/fs_tests.rs (hunk 6 above)
Comment (Anway-Agte) [reply src/tests/fs_tests.rs]: Done!

File: src/tests/fs_tests.rs (hunk 7 above)
Comment (Anway-Agte) [reply src/tests/fs_tests.rs:4046]: Thanks for pointing that out - have added the test!

**Review by namanlalitnyu — APPROVED** [review]
LGTM!

**Review by ve1nard — APPROVED** [review]
lgtm

**Review by yashaswi2000 — APPROVED** [review]
LGTM

File: src/safeposix/syscalls/net_calls.rs [hunk 8]
```diff
@@ -4265,4 +4265,4 @@ impl Cage {
             return syscall_error(Errno::EOPNOTSUPP, "getifaddrs", "invalid ifaddrs length");
         }
     }
-}
+}
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: add newline at end

File: src/tests/networking_tests.rs [hunk 9]
```diff
@@ -4693,4 +4693,4 @@ pub mod net_tests {
         assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
         lindrustfinalize();
     }
-}
+}
```
Comment (rennergade) [inline src/tests/networking_tests.rs]: newline

**Review by rennergade — CHANGES_REQUESTED** [review]
Minor changes requested. But the bigger thing is you haven't done much actual error testing. I suggest adding a test that intentionally triggers errnos.

Also did you say there was a bug in this syscall before or am I misremembering?

**Review by rennergade — APPROVED** [review]

**Pull request merged**